### PR TITLE
Use same gob encoder/decoder for peers and peer connections

### PR DESCRIPTION
**Pull request by awh**
Partially addresses #519 by merging the first two layers of gob.

### CODE REVIEW

File: router/peers.go [hunk 1]
```diff
@@ -25,6 +25,13 @@ type NameCollisionError struct {
 
 type PeerNameSet map[PeerName]struct{}
 
+type ConnectionView struct {
```
Comment (awh) [inline router/peers.go]: Would welcome suggestions for a better name... `GobConnection`?

Comment (rade) [inline router/peers.go]: `ConnectionData`?

File: router/peers.go [hunk 2]
```diff
@@ -233,7 +237,7 @@ func (peers *Peers) decodeUpdate(update []byte) (newPeers map[PeerName]*Peer, de
 	return
 }
 
-func (peers *Peers) applyUpdate(decodedUpdate []*Peer, decodedConns [][]byte) map[PeerName]*Peer {
+func (peers *Peers) applyUpdate(decodedUpdate []*Peer, decodedConns [][]ConnectionView) map[PeerName]*Peer {
 	newUpdate := make(map[PeerName]*Peer)
 	for idx, newPeer := range decodedUpdate {
 		connsBuf := decodedConns[idx]
```
Comment (rade) [inline router/peers.go]: `connsBuf` needs renaming.

File: router/peers.go [hunk 3]
```diff
@@ -298,48 +304,19 @@ func decodePeerNoConns(dec *gob.Decoder) (nameByte []byte, nickName string, uid
...
-		var established bool
-		if err := dec.Decode(&established); err != nil {
-			return err
-		}
-		fun(nameByte, foundAt, outbound, established)
-	}
-}
-
-func readConnsMap(peer *Peer, buf []byte, table map[PeerName]*Peer) map[PeerName]Connection {
+func readConnsMap(peer *Peer, connViews []ConnectionView, table map[PeerName]*Peer) map[PeerName]Connection {
```
Comment (rade) [inline router/peers.go]: function name needs updating; we are not really 'reading' anything anymore here.

File: router/peers.go [hunk 4]
```diff
@@ -210,21 +217,18 @@ func (peers *Peers) decodeUpdate(update []byte) (newPeers map[PeerName]*Peer, de
...
-				return
+				continue
 			}
 			if _, found := peers.table[remoteName]; found {
-				return
+				continue
 			}
 			// Update refers to a peer which we have no knowledge
 			// of. Thus we can't apply the update. Abort.
 			err = UnknownPeerError{remoteName}
```
Comment (rade) [inline router/peers.go]: We should `return` here, and get rid of the `return` below.

**Comment by rade** [comment]
I am curious whether this change makes any difference to the time it takes `NUM_WEAVES=50 bin/multiweave launch --connlimit 100` to settle. (NB: make sure you bump the kernel setting as described in bin/multiweave).

File: router/peers.go (hunk 1 above)
Comment (bboreham) [inline router/peers.go]: Fowler calls this kind of thing a "Data Transfer Object"; if we adopted that pattern then `ConnectionDTO`.  `ConnectionData` not too bad but "Data" is not really adding much to anyone's understanding.

Comment (rade) [inline router/peers.go]: Ah yes. I knew these things had a name. `ConnectionDTO` is good.

Comment (dpw) [inline router/peers.go]: "DTO".  Yuck.

ConnectionUpdate?  ConnectionReport?

Comment (awh) [inline router/peers.go]: `ConnectionSummary`?

**Comment by awh** [comment]
I've made the suggested changes as well as introducing `PeerSummary` to tidy up peer encoding/decoding...

**Comment by rade** [comment]
The linter is moaning.

And the inclusion of `ConnectionSummary` in `PeerSummary` needs further-reaching changes to make sense, so I suggest leaving that out.

**Comment by rade** [comment]
This has made a massive improvement to cluster formation. Running `NUM_WEAVES=50 bin/multiweave launch --connlimit=100` and measuring the time it takes from the first connection to quiescence, I get ~7 minutes on master, and ~2 minutes in this branch.

**Pull request merged**